thewalrus/loop_hafnian_batch.py

```diff
@@ -0,0 +1,301 @@
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Sampling algorithms
+===================
+
+**Module name:** :mod:`thewalrus.loop_hafnian_batch`
+
+.. currentmodule:: thewalrus.loop_hafnian_batch
+
+Tools to evaluate the loop hafnian batch to do gaussian boson sampling.
+
+Summary
+-------
+
+.. autosummary::
+    _calc_loop_hafnian_batch_even
+    _calc_loop_hafnian_batch_odd
+    add_batch_edges_even
+    add_batch_edges_odd
+    loop_hafnian_batch
+
+Code details
+------------
+"""
+import numpy as np
+import numba
+from thewalrus._hafnian import (
+    precompute_binoms,
+    matched_reps,
+    find_kept_edges,
+    f_loop,
+    f_loop_odd,
+    get_submatrices,
+    get_submatrix_batch_odd0,
+)
+
+# pylint: disable = too-many-arguments, not-an-iterable
+@numba.jit(nopython=True, parallel=True, cache=True)
+def _calc_loop_hafnian_batch_even(
+    A, D, fixed_edge_reps, batch_max, odd_cutoff, glynn=True
+):  # pragma: no cover
+    r"""Calculate the batched loop hafnian for paired mode.
+
+    Args:
+        A (array): input matrix.
+        D (array): vector to find loop hafnian.
+        fixed_edge_reps (list): fixed numbers of edge repetition.
+        batch_max (array): maximum number of photons for m mode.
+        odd_cutoff (array): cutoff for odd modes.
+        glynn (boolean): determines the method used to evaluate the loop hafnian batch.
+
+    Returns:
+        H_batch (array): vector giving loop hafnian for every photon number.
+    """
+    oddloop = D[0]
+    oddV = A[0, :]
+
+    n = A.shape[0]
+    N_fixed = 2 * fixed_edge_reps.sum()  # number of photons
+
+    N_max = N_fixed + 2 * batch_max + odd_cutoff
+
+    edge_reps = np.concatenate((np.array([batch_max]), fixed_edge_reps))
+    steps = np.prod(edge_reps + 1)
+    # precompute binomial coefficients
+    max_binom = edge_reps.max() + odd_cutoff
+    binoms = precompute_binoms(max_binom)
+
+    H_batch = np.zeros(2 * batch_max + odd_cutoff + 1, dtype=np.complex128)
+    # prange to range
+    for j in numba.prange(steps):
+
+        Hnew = np.zeros(2 * batch_max + odd_cutoff + 1, dtype=np.complex128)
+
+        kept_edges = find_kept_edges(j, edge_reps)
+        edges_sum = kept_edges.sum()
+
+        binom_prod = 1.0
+        for i in range(1, n // 2):
+            binom_prod *= binoms[edge_reps[i], kept_edges[i]]
+
+        if glynn:
+            delta = 2 * kept_edges - edge_reps
+        else:
+            delta = kept_edges
```

Comment:
a one-liner for this:
```suggestion
        delta = 2 * kept_edges - edge_reps if glynn else kept_edges
```
though I'm not sure if this is something of the likes of numba.